File: 3-LongestSubstringWithoutRepeatingCharacters/3-LongestSubstringWithoutRepeatingCharacters.py
# Last updated: 2/11/2026, 11:08:51 AM
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """

        #use left and right, where left is index 0 and right is at index 1
        #we check for duplicate characters at each window restricted by left and right

        #use a set

        #loop
        
        #if within the window, a duplicate is found, we move left to the right, until there are no duplicates

        #once there are no duplicates move right to the right



        left = 0
        right = 1
        maxLen = 0
        flag = False

        while (right <= len(s)):
            st = set()

            for ch in s[left:right]:
                if ch not in st:
                    flag = False
                    st.add(ch)
                else:
                    flag = True
                    break

            if flag:
                left+=1
            else:
                maxLen = max(maxLen, len(st))
                right+=1

        return maxLen

File: 3-LongestSubstringWithoutRepeatingCharacters/3-LongestSubstringWithoutRepeatingCharacters.java
// Last updated: 2/11/2026, 11:08:53 AM
class Solution {
    public int lengthOfLongestSubstring(String s) {

        int windowStart = 0;
        int max = 0;
        HashSet<Character> set = new HashSet<Character>();

        for (int windowEnd = 0; windowEnd < s.length(); windowEnd++){
            char temp = s.charAt(windowEnd);
            if(!set.contains(temp))
            {
                set.add(temp);
            }
            else{
                while(set.contains(temp)){
                    set.remove(s.charAt(windowStart));
                    windowStart++;
                }
                set.add(temp);
            }
            if (max<set.size()){
                    max = set.size();
                }
        }
        return max;    
    }
}